### PR TITLE
Racing Section

### PR DESCRIPTION
**Pull request by topepo**
A pretty straight-forward section on racing. One hands-on at the end.

### CODE REVIEW

**Comment by EmilHvitfeldt** [comment]
Is it on purpose that you pushing this into main instead of 2023-NYR?

**Comment by topepo** [comment]
Nope. I forgot that main was the default. Should be fixed now.

File: slides/advanced-02-racing.qmd [hunk 1]
```diff
@@ -0,0 +1,349 @@
...
+
+
+```{r}
+#| label: plot-race
+#| out-width: 100%
+#| fig-align: center
+#| fig-width: 5
+#| fig-height: 3.8
+
+plot_race(lgbm_race_res)
```
Comment (simonpcouch) [inline slides/advanced-02-racing.qmd:330]: Is the "Analysis Stage" the number of resamples evaluated before the model was no longer evaluated? If so, ideally x axis breaks would be integers.

**Review by simonpcouch — COMMENTED** [review]
This is in a good spot!

File: slides/advanced-02-racing.qmd [hunk 2]
```diff
@@ -0,0 +1,349 @@
...
+Both of these methods can lead to enormous speed-ups. 
+
+
+## Model Racing 
+
+[_Racing_](https://scholar.google.com/scholar?hl=en&as_sdt=0%2C7&q=+Hoeffding+racing) is an old tool that we can use to go even faster. 
+
+1. Evaluate all of the candidate models but only for a few resamples. 
+1. Determine which candidates have a low probability of being selected.
+1. Eliminate poor candidates
```
Comment (hfrick) [inline slides/advanced-02-racing.qmd:139]:
```suggestion
1. Eliminate poor candidates.
```

File: slides/advanced-02-racing.qmd [hunk 3]
```diff
@@ -0,0 +1,349 @@
...
+1. Repeat with next resample. 
+
+This can result in fitting a small number of models. 
+
+
+## Discarding Candidates
+
+How to we eliminate tuning parameter combinations?
```
[the remark this answers is not in the document]
Comment (hfrick) [reply slides/advanced-02-racing.qmd:147]: This switches between "candidate models" and "tuning parameter combinations". Not sure if we care to only use one term consistently?

File: slides/advanced-02-racing.qmd [hunk 4]
```diff
@@ -0,0 +1,349 @@
...
+ - or a t-test on their differences matched by resamples
+
+With $n = 10$ resamples, the confidence interval is (`r signif(race_ci$lower[9], 2)`, `r signif(race_ci$upper[9], 2)`), indicating that candidate number 2 has smaller error. 
+
+What if we were to compare each model candidate to the current best at each resample? 
+
+One shows superiority when `r which.max(race_ci$lower > 0)` resamples have been evaluated.
+
+
+## Evaluating Differences in Candidates
```
Comment (hfrick) [inline slides/advanced-02-racing.qmd:252]: I liked how the plots in your conf talk from 2021 showed the process per iteration/resample, i.e. that "evaluating the differences between candidates" and "candidates are being discarded" were being shown together.

For anyone not Max: from about the 5:30 mark of https://posit.co/resources/videos/whats-new-in-tidymodels/

**Review by hfrick — COMMENTED** [review]
Always like the racing stuff! 🏎️

File: slides/advanced-02-racing.qmd [hunk 5]
```diff
@@ -0,0 +1,349 @@
...
+hotel_rates <- 
+  data_hotel_rates() %>% 
+  sample_n(5000) %>% 
+  arrange(arrival_date) %>% 
+  select(-arrival_date_num, -arrival_date) %>% 
+  mutate(
+    company = factor(as.character(company)),
+    country = factor(as.character(country)),
+    agent = factor(as.character(agent))
+  )
```
Comment (EmilHvitfeldt) [inline slides/advanced-02-racing.qmd:78]: Are you planning on commenting on `factor(as.character(`? It might be non-obvious why you are doing it 😄

File: slides/advanced-02-racing.qmd [hunk 6]
```diff
@@ -0,0 +1,349 @@
...
+lgbm_spec <- 
+  boost_tree(trees = tune(), learn_rate = tune(), mtry = tune(), 
+             min_n = tune()) %>% 
+  set_mode("regression") %>% 
+  set_engine("lightgbm")
+
+lgbm_wflow <- workflow(hotel_rec, lgbm_spec)
+```
+
+## Making Grid Search More Effcieint
```
Comment (EmilHvitfeldt) [inline slides/advanced-02-racing.qmd:120]:
```suggestion
## Making Grid Search More Efficient
```

File: slides/advanced-02-racing.qmd [hunk 7]
```diff
@@ -0,0 +1,349 @@
...
+
+
+## Model Racing 
+
+[_Racing_](https://scholar.google.com/scholar?hl=en&as_sdt=0%2C7&q=+Hoeffding+racing) is an old tool that we can use to go even faster. 
+
+1. Evaluate all of the candidate models but only for a few resamples. 
+1. Determine which candidates have a low probability of being selected.
+1. Eliminate poor candidates
+1. Repeat with next resample.
```
Comment (EmilHvitfeldt) [inline slides/advanced-02-racing.qmd:140]: I feel this needs a 5th step to determine how/when it stops.

File: slides/advanced-02-racing.qmd [hunk 8]
```diff
@@ -0,0 +1,349 @@
...
+
+race_ci %>% 
+  ggplot(aes(`number of resamples`, difference)) + 
+  geom_point() +
+  geom_hline(yintercept = 0, col = "red", lty = 2) +
+  geom_errorbar(aes(ymin = lower, ymax = upper), width = 1 / 10, alpha = 0.5) +
+  labs(y = "error difference from best")
+```
+
+## Interim Analysis of Results
```
Comment (EmilHvitfeldt) [inline slides/advanced-02-racing.qmd:271]: I would like a `<br>` or two in this slide if possible. it is a little wall-of-text-y

File: slides/advanced-02-racing.qmd [hunk 9]
```diff
@@ -0,0 +1,349 @@
...
+
+One version of racing uses a _mixed model ANOVA_ to construct one-sided confidence intervals for each candidate versus the current best. 
+
+Any candidates whose bound does not include zero are discarded. 
+
+The resamples are analyzed in a random order. 
+
+[Kuhn (2014)](https://arxiv.org/abs/1405.6974) has examples and simulations to show that the method works. 
+
+The finetune package has functions `tune_race_anova()` and `tune_race_win_loss()`.
```
Comment (EmilHvitfeldt) [inline slides/advanced-02-racing.qmd:281]:
```suggestion
The [finetune](https://finetune.tidymodels.org/) package has functions `tune_race_anova()` and `tune_race_win_loss()`. 
```

**Review by EmilHvitfeldt — COMMENTED** [review]
Great set of slides!

missing `hexes()` 😄

File: slides/advanced-02-racing.qmd [hunk 10]
```diff
@@ -0,0 +1,349 @@
...
+This can result in fitting a small number of models. 
+
+
+## Discarding Candidates
+
+How to we eliminate tuning parameter combinations? 
+
+There are a few methods to do so. We'll use one based on analysis of variance (ANOVA). 
+
+_However_... there is typically a large effect due to resamples in the data.
```
[the remark this answers is not in the document]
Comment (topepo) [reply slides/advanced-02-racing.qmd:151]: Changed it to:

> there is typically a large difference between resamples in the results.

File: slides/advanced-02-racing.qmd (hunk 1 above)
Comment (topepo) [reply slides/advanced-02-racing.qmd:330]: I agree. I'll put an issue in to finetune for that. I added a `scale_x_continuous(breaks= pretty_breaks())`

File: slides/advanced-02-racing.qmd (hunk 4 above)
Comment (topepo) [reply slides/advanced-02-racing.qmd:252]: I'll add an issue to make this same animation for the posit conf version. (not sure that I have enough time currently)

File: slides/advanced-02-racing.qmd (hunk 5 above)
Comment (topepo) [reply slides/advanced-02-racing.qmd:78]: I will go back and add it to the initial discussion of the data
